provider/aws Add aws_vpc_endpoint_service data source

ewbankkit opened the pull request:
Add `aws_vpc_endpoint_services` data source to provide a list of all supported AWS services that can be specified when creating a VPC endpoint.
Add documentation.
Add acceptance test:
```
$ make testacc TEST=./builtin/providers/aws TESTARGS='-run=TestAccDataSourceAwsVpcEndpointServices'
```

ewbankkit commented:
Ideally I'd like to be able to do this:
```
# Declare the data source
data "aws_vpc_endpoint_services" "es" {}

# Create a VPC
resource "aws_vpc" "foo" {
    cidr_block = "10.0.0.0/16"
}

# Create all VPC endpoints
resource "aws_vpc_endpoint" "ep" {
    count = "${length(data.aws_vpc_endpoint_services.es.names)}"
    vpc_id = "${aws_vpc.foo.id}"
    service_name = "${data.aws_vpc_endpoint_services.es.names[count.index]}"
}
```
But I get the error:
```
* aws_vpc_endpoint.ep: resource count can't reference resource variable: data.aws_vpc_endpoint_services.es.names
```
That requires progress on https://github.com/hashicorp/terraform/issues/1497 and https://github.com/hashicorp/terraform/pull/8521.

ewbankkit commented:
Tempted to add another data source `aws_vpc_endpoint_service` (singular) that has an attribute `service` with a value such as `s3`. It would return the service name for that service:
```
provider "aws" {
  region = "us-west-2"
}

data "aws_vpc_endpoint_service" "s3" {
  service = "s3"
}
```
Then `${data.aws_vpc_endpoint_service.s3}` has the value `"com.amazonaws.us-west-2.s3"`.

kwilczynski commented on builtin/providers/aws/data_source_aws_vpc_endpoint_services.go:
```diff
@@ -0,0 +1,47 @@
+	services := aws.StringValueSlice(resp.ServiceNames)
+
+	if err := d.Set("names", services); err != nil {
+		return fmt.Errorf("[WARN] Error setting VPC Endpoint Services: %s", err)
```
So, either an error or a warning? 😄

kwilczynski commented on builtin/providers/aws/data_source_aws_vpc_endpoint_services_test.go:
```diff
@@ -0,0 +1,58 @@
+			return err
+		}
+		if n < 1 {
+			return fmt.Errorf("Number of services seem suspiciously low: %d", n)
```
That is a very polite error message. I would just spell it out as it is more bluntly. Hehe.

ewbankkit commented:
The more I though about the concept of return a list of all VPC endpoint service names (even though there is only one right now, s3), the less I liked it. There seems to be no good way of ensuring that the ordering in the list stays consistent when new services are added.
So, I renamed the resource to `aws_vpc_endpoint_service`, added an attribute `service` which selects the service and renamed the exported attribute to `service_name`.
Usage would then be like:
```
# Declare the data source
data "aws_vpc_endpoint_service" "s3" {
  service = "s3"
}
 
 # Create a VPC
 resource "aws_vpc" "foo" {
     cidr_block = "10.0.0.0/16"
 }
 
 # Create a VPC endpoint
 resource "aws_vpc_endpoint" "s3" {
     vpc_id = "${aws_vpc.foo.id}"
     service_name = "${data.aws_vpc_endpoint_service.s3.service_name}"
 }
```

Acceptance test becomes:
```
$ make testacc TEST=./builtin/providers/aws TESTARGS='-run=TestAccDataSourceAwsVpcEndpointService'
```

stack72 commented:
Hi @ewbankkit 

what is the status of this PR - still a WIP? 

P.

ewbankkit commented:
Going to do a quick once-over and then "un-WIP" it.

stack72 commented:
👍 :)

ewbankkit commented:
Rebased and removed `[WIP]` prefix.
Ready for consideration.

stack72 commented:
Hi @ewbankkit 

Thanks for the work here - this data_source is going to be pretty limited for a while but I don't think that should stop us for now :)

```
% make testacc TEST=./builtin/providers/aws TESTARGS='-run=TestAccDataSourceAwsVpcEndpointService'
==> Checking that code complies with gofmt requirements...
go generate $(go list ./... | grep -v /terraform/vendor/)
2016/12/13 10:39:29 Generated command/internal_plugin_list.go
TF_ACC=1 go test ./builtin/providers/aws -v -run=TestAccDataSourceAwsVpcEndpointService -timeout 120m
=== RUN   TestAccDataSourceAwsVpcEndpointService
--- PASS: TestAccDataSourceAwsVpcEndpointService (13.27s)
PASS
ok  	github.com/hashicorp/terraform/builtin/providers/aws	13.299s
```

Thanks

Paul
